File: 189. Rotate Array/189.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        // when supposed to do nothing reversing it will cause error
        if (nums.size() == 1 || k == 0)
            return;

        // finding pattern
        int steps = k % nums.size();
        
        if (steps == 0)
            return;
        
        reverse(nums.begin(), nums.end());
        reverse(nums.begin(), nums.begin()+steps);
        reverse(nums.begin()+steps, nums.end());
        
    }
};

File: 189. Rotate Array/189.py
"""
intuition: nums = nums[len(nums)-k:] + nums[0: len(nums) - k - 1]
divide and conquer, reversing list
time: O(n)
space: O(1)
"""
class Solution:

    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        k %= len(nums)
        nums.reverse()
        nums[0:k] = reversed(nums[0:k])
        nums[k:] = reversed(nums[k:])

        
